src/clj/ataru/organization_service/session_organizations.clj

```diff
@@ -1,25 +1,47 @@
 (ns ataru.organization-service.session-organizations
   (:require
-   [schema.core :as s]
-   [ataru.organization-service.organization-client :as organization-client]
-   [ataru.organization-service.user-rights :refer [Right]]))
+    [schema.core :as s]
+    [ataru.organization-service.organization-client :as organization-client]
+    [ataru.organization-service.user-rights :refer [Right]]
+    [ataru.organization-service.organization-service :as organization-service]))
 
-(defn- right-organizations [session] (-> session :identity :user-right-organizations))
+(defn- right-organizations
+  [session]
+  (-> session :identity :user-right-organizations))
 
 (defn- all-org-oids [organization-service organizations]
   (let [all-organizations (.get-all-organizations organization-service organizations)]
     (set (map :oid all-organizations))))
 
+(defn- filter-orgs-for-rights
+  [session rights organizations]
+  (if (-> session :identity :superuser)
+    organizations
+    (filter
+      (fn [org]
+        (-> (-> session :identity :organizations)
+            (get (keyword (:oid org)))
+            :rights
+            (set)
+            (clojure.set/intersection (set rights))
+            (not-empty)))
+      organizations)))
+
 (defn right-seq? [val] (s/validate [Right] val))
 
-(defn select-organizations-for-rights [session rights]
+(defn select-organizations-for-rights [organization-service session rights]
   {:pre [(right-seq? rights)]}
-  (let [right-orgs (right-organizations session)]
-    (->> rights
-         (map #(get right-orgs %))
-         (remove nil?)
-         flatten
-         distinct)))
+  (if-let [selected-organization (:selected-organization session)]
+    (filter-orgs-for-rights
+      session
+      rights
+      (.get-all-organizations organization-service [selected-organization]))
+    (let [right-orgs (right-organizations session)]
+      (->> rights
+           (map #(get right-orgs %))
+           (remove nil?)
+           (flatten)
+           (set)))))
```

Comment:
Näyttäisi siltä, että kaikissa `select-organizations-for-rights` kutsupaikoissa päädytään lopulta tekemään myös `get-all-organizations` kutsu, joten sen voisi upottaa tähän.